File: scripts/port_check.go
package main

import (
	"context"
	"flag"
	"fmt"
	"net"
	"os"
	"strings"
	"sync"
	"time"
)

func main() {
	startTime := time.Now()

	// Define command-line flags
	ports := flag.String("p", "", "Port numbers (comma-separated, e.g., 80,443,3306)")
	help := flag.Bool("h", false, "Show this help message")

	flag.Parse()

	if *help || len(os.Args) == 1 {
		printUsage()
		os.Exit(0)
	}

	if *ports == "" {
		fmt.Println("Error: Port numbers are required. Use -p or --port option.")
		printUsage()
		os.Exit(1)
	}

	target := flag.Arg(0)
	if target == "" {
		fmt.Println("Error: Target is required.")
		printUsage()
		os.Exit(1)
	}

	fmt.Printf("Target: %s\n", target)
	fmt.Printf("Ports: %s\n", *ports)

	// Resolve target and get IPs
	ipAddrs, err := resolveTarget(target)
	if err != nil {
		fmt.Printf("Error resolving target: %v\n", err)
		os.Exit(1)
	}

	// Separate IPs into IPv4 and IPv6 lists
	var ipv4Targets, ipv6Targets []net.IP
	for _, ip := range ipAddrs {
		if ip.To4() != nil {
			ipv4Targets = append(ipv4Targets, ip)
		} else {
			ipv6Targets = append(ipv6Targets, ip)
		}
	}

	// Parse ports
	portsToScan, err := parsePorts(*ports)
	if err != nil {
		fmt.Printf("Error parsing ports: %v\n", err)
		os.Exit(1)
	}

	// Run scans for IPv4 if present
	if len(ipv4Targets) > 0 {
		fmt.Println("\n--- Starting scans for IPv4 ---")
		for _, ip := range ipv4Targets {
			fmt.Printf("Scanning target: %s\n", ip)
			scanAndPrintResultsForIP(ip, portsToScan)
		}
	}

	// Run scans for IPv6 if present
	if len(ipv6Targets) > 0 {
		fmt.Println("\n--- Starting scans for IPv6 ---")
		for _, ip := range ipv6Targets {
			fmt.Printf("Scanning target: %s\n", ip)
			scanAndPrintResultsForIP(ip, portsToScan)
		}
	}

	duration := time.Since(startTime)
	fmt.Fprintf(os.Stderr, "Elapsed time: %.3f seconds.\n", duration.Seconds())
}

// scanAndPrintResultsForIP handles the scanning and ordered printing for a single IP
func scanAndPrintResultsForIP(ip net.IP, portsToScan []string) {
	results := make(chan string)
	var wg sync.WaitGroup

	for _, port := range portsToScan {
		wg.Add(2) // TCP and UDP
		go scanPort(ip.String(), port, "tcp", &wg, results)
		go scanPort(ip.String(), port, "udp", &wg, results)
	}

	// Start a goroutine to close the results channel once all scanning is done
	go func() {
		wg.Wait()
		close(results)
	}()

	// Collect and print results from the channel
	for result := range results {
		fmt.Println(result)
	}
	fmt.Println() // Add a blank line for spacing after each IP block
}

func printUsage() {
	fmt.Println("Usage: go run port_check.go [OPTIONS] <target>")
	fmt.Println("")
	fmt.Println("Options:")
	flag.PrintDefaults()
	fmt.Println("")
	fmt.Println("Targets:")
	fmt.Println("  IPv4 address     Direct IPv4 target")
	fmt.Println("  IPv6 address     Direct IPv6 target")
	fmt.Println("  Domain name       Will resolve to both IPv4 and IPv6 automatically")
	fmt.Println("")
	fmt.Println("Examples:")
	fmt.Println("  go run port_check.go -p 80,443 example.com")
	fmt.Println("  go run port_check.go -p 80,443 1.1.1.1")
}

func resolveTarget(target string) ([]net.IP, error) {
	if ip := net.ParseIP(target); ip != nil {
		return []net.IP{ip}, nil
	}

	var allIPs []net.IP
	resolver := net.Resolver{}

	// Explicitly look up IPv4 addresses
	ipv4s, err4 := resolver.LookupIP(context.Background(), "ip4", target)
	if err4 == nil {
		allIPs = append(allIPs, ipv4s...)
	}

	// Explicitly look up IPv6 addresses
	ipv6s, err6 := resolver.LookupIP(context.Background(), "ip6", target)
	if err6 == nil {
		allIPs = append(allIPs, ipv6s...)
	}

	// If both lookups failed, return an error
	if err4 != nil && err6 != nil {
		return nil, fmt.Errorf("could not resolve domain %s: %v", target, err4)
	}

	if len(allIPs) == 0 {
		return nil, fmt.Errorf("no IP addresses found for domain %s", target)
	}

	return allIPs, nil
}

func parsePorts(portStr string) ([]string, error) {
	ports := strings.Split(portStr, ",")
	if len(ports) == 0 {
		return nil, fmt.Errorf("no ports specified")
	}
	return ports, nil
}

// scanPort attempts to connect to a specific port and sends the result to a channel.
// Note: UDP scanning with this method is not always reliable. A timeout does not
// definitively mean a port is closed, as open UDP ports are not required to respond.
// The 'nmap' output 'open|filtered' is a more accurate representation of this uncertainty.
func scanPort(ip string, port string, protocol string, wg *sync.WaitGroup, results chan<- string) {
	defer wg.Done()
	// For IPv6, the address must be enclosed in square brackets.
	address := net.JoinHostPort(ip, port)
	timeout := 2 * time.Second
	conn, err := net.DialTimeout(protocol, address, timeout)

	if err != nil {
		// Do nothing if port is closed
		return
	}

	conn.Close()
	results <- fmt.Sprintf("% -5s % -10s %s", port, strings.ToUpper(protocol), "Open")
}

File: scripts/port_check.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.11"
# dependencies = []
# ///
"""Check for open TCP/UDP ports on a target host.

Runnable with uv (no external tools needed):
    uv run port_check.py -p 80,443 example.com
    uv run port_check.py --port 80-90 2606:4700:4700::1111

This is a pure-Python stdlib port of port_check.sh / port_check.go.
Note: UDP scanning via connect is unreliable; a timeout does not
definitively mean a port is closed (open UDP ports may not respond).
"""

from __future__ import annotations

import socket
import sys
import time
from concurrent.futures import ThreadPoolExecutor, as_completed

DEFAULTS = {"timeout": 2, "workers": 100}


class PortChecker:
    def __init__(self) -> None:
        self.ports: list[tuple[int, int]] = []
        self.target = ""

    def parse_ports(self, port_str: str) -> None:
        for part in port_str.split(","):
            part = part.strip()
            if not part:
                return
            if "-" in part:
                start_s, end_s = part.split("-", 1)
                start, end = int(start_s), int(end_s)
            else:
                start = end = int(part)
            if start < 1 or end > 65535 or start > end:
                sys.exit(f"Error: Invalid port format: {part} (must be 1-65535).")
            self.ports.append((start, end))

    def scan_port(self, ip: str, port: int, proto: str) -> None:
        family = socket.AF_INET6 if ":" in ip else socket.AF_INET
        sock_type = socket.SOCK_STREAM if proto == "tcp" else socket.SOCK_DGRAM
        sock = socket.socket(family, sock_type)
        sock.settimeout(self.timeout)
        try:
            sock.connect((ip, port))
        except OSError:
            return
        finally:
            sock.close()
        print("{:<5} {:<10} Open".format(str(port), proto.upper()))

    def scan_ip(self, ip: str) -> None:
        jobs: list[tuple[str, int, str]] = []
        for start, end in self.ports:
            for port in range(start, end + 1):
                jobs.append((ip, port, "tcp"))
                jobs.append((ip, port, "udp"))
        with ThreadPoolExecutor(max_workers=self.workers) as pool:
            futures = {pool.submit(self.scan_port, *job): job for job in jobs}
            for _ in as_completed(futures):
                pass

    def resolve_target(self) -> list[str]:
        try:
            socket.inet_pton(socket.AF_INET6, self.target)
            return [self.target]
        except OSError:
            pass
        try:
            socket.inet_pton(socket.AF_INET, self.target)
            return [self.target]
        except OSError:
            pass
        try:
            return list(dict.fromkeys(
                info[4][0] for info in socket.getaddrinfo(self.target, None)
            ))
        except socket.gaierror as exc:
            sys.exit(f"Error resolving target {self.target}: {exc}")

    def build_argv(self, argv: list[str]) -> None:
        if len(argv) == 1 or argv[0] in ("-h", "--help"):
            self.usage()
            sys.exit(0)
        if argv[0] in ("-p", "--port"):
            self.parse_ports(argv[1])
            if len(argv) >= 3:
                self.target = argv[2]
        else:
            self.target = argv[0]
            for k, val in enumerate(argv):
                if val in ("-p", "--port") and k + 1 < len(argv):
                    self.parse_ports(argv[k + 1])
        if not self.ports:
            sys.exit("Error: Port numbers are required. Use -p or --port option.")
        if not self.target:
            sys.exit("Error: Target is required.")
        self.timeout = DEFAULTS["timeout"]
        self.workers = DEFAULTS["workers"]

    def usage(self) -> None:
        print(
            """Usage: uv run port_check.py [OPTIONS] <target>

Options:
  -p, --port       Port numbers or ranges (comma-separated, e.g., 80,443,1-1024)
  -h, --help       Show this help message

Targets:
  IPv4 address     Direct IPv4 target
  IPv6 address     Direct IPv6 target
  Domain name      Will resolve to both IPv4 and IPv6 automatically

Examples:
  uv run port_check.py -p 80,443 example.com
  uv run port_check.py --port 80,443 1.1.1.1
  uv run port_check.py -p 22,80,443 2606:4700:4700::1111"""
        )

    def run(self) -> None:
        ips = self.resolve_target()
        if not ips:
            sys.exit(f"Error: Could not resolve target {self.target}.")
        for ip in ips:
            print(f"\n--- Starting scans for {ip} ---")
            self.scan_ip(ip)


def main(argv: list[str]) -> int:
    checker = PortChecker()
    checker.build_argv(argv)
    checker.run()
    return 0


if __name__ == "__main__":
    start = time.time()
    code = main(sys.argv[1:])
    print(f"Elapsed time: {time.time() - start:.3f} seconds.", file=sys.stderr)
    sys.exit(code)
